index.d.ts

```diff
@@ -30,33 +31,103 @@ declare type ReturnTypeOf<T extends AnyFunction | AnyFunction[]> =
     ? UnionToIntersection<Exclude<ReturnType<T[number]>, void>>
     : never;
 
+type ClassWithPlugins = Constructor<any> & {
+  plugins: any[];
+};
+
+type ConstructorRequiringVersion<Class extends ClassWithPlugins, PredefinedOptions> = {
+  defaultOptions: PredefinedOptions;
+} & (PredefinedOptions extends { version: string }
+  ? {
+      new <NowProvided>(options?: NowProvided): Class & {
+        options: NowProvided & PredefinedOptions;
+      };
+    }
+  : {
+      new <NowProvided>(options: Base.Options & NowProvided): Class & {
+        options: NowProvided & PredefinedOptions;
+      };
+    });
+
 export declare class Base<TOptions extends Base.Options = Base.Options> {
   static plugins: Plugin[];
+
+  /**
+   * Pass one or multiple plugin functions to extend the `Base` class.
+   * The instance of the new class will be extended with any keys returned by the passed plugins.
+   * Pass one argument per plugin function.
+   *
+   * ```js
+   * export function helloWorld() {
+   *   return {
+   *     helloWorld () {
+   *       console.log('Hello world!');
+   *     }
+   *   };
+   * }
+   *
+   * const MyBase = Base.plugin(helloWorld);
+   * const base = new MyBase();
+   * base.helloWorld(); // `base.helloWorld` is typed as function
+   * ```
+   */
   static plugin<
-    S extends Constructor<any> & {
-      plugins: any[];
-    },
-    T1 extends Plugin,
-    T2 extends Plugin[]
+    Class extends ClassWithPlugins,
+    Plugins extends [Plugin, ...Plugin[]],
```

Comment:
New little nifty improvement recently added to tuples in TypeScript ✨ no more multiple parameters!